### PR TITLE
RingBuf split_off and append impls

### PR DESCRIPTION
**Pull request by Gankra**
Working on just knocking these out for all the collections so that there's something there.

### CODE REVIEW

**Comment by Gankro** [comment]
Found an error, retesting

**Comment by Gankro** [comment]
All fixed.

File: src/libcollections/ring_buf.rs [hunk 1]
```diff
@@ -1289,6 +1289,93 @@ impl<T> RingBuf<T> {
...
+        let other_len = len - at;
+        let mut other = RingBuf::with_capacity(other_len);
+
+        unsafe {
+            let (first_half, second_half) = self.as_slices();
+
+            let first_len = first_half.len();
+            let second_len = second_half.len();
+            if at < first_len {
+                // `at` lies in the first half.
```
Comment (huonw) [inline src/libcollections/ring_buf.rs]: You're explicitly not counting `at == first_len` as being in the first half?

Comment (Gankro) [inline src/libcollections/ring_buf.rs]: if my ringbuf looks like `[0, 1, 2, 3][4, 5, 6]` and `at = 4`, then `at == first_len` and `at` is in the second half. More importantly: I don't want *any* of the elements in the first half.

Comment (huonw) [inline src/libcollections/ring_buf.rs]: Well, either branch works if `at == first_len`, but yeah, this approach is better.

**Comment by Gankro** [comment]
Crap, I accidentally force pushed an old version of this branch on my compile station. Will fix when I get home.

**Pull request merged**